### DISCUSSION TITLE
Email configuration problem

### DISCUSSION THREAD

**Discussion by lmsalgado** [consuldemocracy/consuldemocracy#5822] [Q&A]
I have read everything I could find, but cannot make the emails to be sent.

Accordingly with the instructions on https://github.com/consuldemocracy/installer (and several other places), I configured my email account in the file /home/deploy/consul/shared/config/secrets.yml.

I tried with a gmail account (with 2-step authentication enabled and an app password):
# secret_key_base: ""
  server_name: "my_domain.com"
  time_zone: "Europe/Lisbon"
  mailer_delivery_method: :smtp
  smtp_settings:
    :address:              "smtp.gmail.com"
    :port:                 587
    :domain:               "my_domain.com "
    :user_name:            "my_account@gmail.com"
    :password:             "my_app_password"
    :authentication:       "plain"
    :enable_starttls_auto: true

Also tried with a free SMTP Server (MailerSend):
# secret_key_base: ""
  server_name: "my_domain.com"
  time_zone: "Europe/Lisbon"
  mailer_delivery_method: :smtp
  smtp_settings:
    :address:              "smtp.mailersend.net"
    :port:                 587
    :domain:               "my_domain.com "
    :user_name:            "my_account.mlsender.net"
    :password:             "my_password"
    :authentication:       "plain"
    :enable_starttls_auto: true

The file secrets.yml repeats this configurations 3 different times, for staging, preproduction and production, I applied the same configurations for all 3. The rest of the file I left with the default settings.
Also changed the site on Settings - Configuration settings - Sender email address, using the same user_name as applied in secrets.yml.

Still emails are not getting out, neither an error is raised on the site.  
Am I missing something? What can I do to try to solve this?

**Comment by javierm** [comment]
Hi, @lmsalgado :smile:.

Just to be sure, are you restarting the application and the delayed job background processes after changing the secrets file? :thinking:

Assuming you're using a production environment, to restart the application, run the following command on the server:

```bash
systemctl --user restart puma_consul_production
```

To restart the background processes:

```bash
RAILS_ENV=production bundle exec bin/delayed_job -m -n 2 restart
```

If you're using capistrano to deploy your code, you can run both commands at the same time by running the following command _on your local machine_:

```bash
cap production deploy:restart
```

If restarting the server doesn't solve the issue, could you check whether there are errors in the logs (particularly in the `log/delayed_job.log` file)?

**Comment by lmsalgado** [comment]
Hi @javierm, thank you so much for your help.

The logs show absolutely nothing about it. If I try to send email with simple python script it works correctly. It seems to me the emails are being inserted in the asynchronous queue but then not sent.
Not using capistrano.

Oddly, both your commands are not working:

systemctl --user restart puma_consul_production
Failed to restart puma_consul_production.service: Unit puma_consul_production.service not found.


RAILS_ENV=production bundle exec bin/delayed_job -m -n 2 restart
Could not find pg-1.5.8, puma-5.6.9, redcarpet-3.6.0, rinku-2.0.6, json-2.7.2, bcrypt-3.1.20, sassc-2.4.0, nio4r-2.7.3, websocket-driver-0.7.6, rbtree3-0.7.1, racc-1.8.1, nkf-0.2.0, bigdecimal-3.1.8, date-3.3.4 in locally installed gems
Run `bundle install` to install missing gems.

(Note: I did try to run "bundle install" and it messed completely the installation, bringing the site down, to the point that I had to recover last night backup)

**Reply by javierm** [reply]
Hi, @lmsalgado :smile:.

```
systemctl --user restart puma_consul_production
Failed to restart puma_consul_production.service: Unit puma_consul_production.service not found.
```

Are you using the production environment? What are the contents of the `~/.config/systemd/user/` folder?

> I did try to run "bundle install" and it messed completely the installation

I know it's too late, but don't do it. Use the capistrano tasks to do so, although, if the application was running, chances are it wasn't necessary :thinking:.

**Reply by lmsalgado** [reply]
> Are you using the production environment? 


I think so, but I did nothing to change the environment, I run the steps on CONSUL DEMOCRACY Installer and start using (no capistrano). On secrets.yml I changed only the smtp configs for production (and after install the emails are working until I reboot the server)

>  What are the contents of the ~/.config/systemd/user/ folder?

I do not have those folders on ~, I have them on /home/deploy/.config/systemd/user

```
ls /home/deploy/.config/systemd/user
default.target.wants  puma_consul_production.service  puma_consul_production.socket  sockets.target.wants
```

**Reply by javierm** [reply]
Oh, sorry :pray:. Forgot to say that all these commands I'm mentioning should be run using the `deploy` user (if you're using a different user, that might be the reason why they were failing).

**Reply by lmsalgado** [reply]
Ohh that explain a lot, I have been using root...   :)   Thank you, I won't make the same mistake again

**Comment by lmsalgado** [comment]
Adding some info about the last error:
The path on consul/current/.bundle/config is pointing to consul/shared/bundle
all the gems not found on the error message, are in consul/shared/bundle/ruby/3.2.0/gems

Instead of restarting the processes, I have been rebooting the whole server (and still not emails sent).
Apart from these commands returning errors and the emails not sent, the application is working fine.

**Comment by lmsalgado** [comment]
An update:

If I "reinstal" Consul with the comand: "ansible-playbook -v consul.yml -i hosts"  all the email on the queue are sent. Furthermore, new emails are working fine (for instance: resend account confirmation email) **until the next server reboot**. If I reboot the server the emails are queued (I can see them in the database, in the table delayed_jobs) but never sent neither marked as failed.

So, configurations are ok, and apparently all needed dependencies are there. It seems that some service or path is not restarted after reboot.
How can i find out what is happening?

**Reply by javierm** [reply]
OK, a couple of questions.

First, which version of Consul Democracy are you using? There was a bug in one specific version that might explain this issue.

Second, could you paste the contents of your `delayed_job.log` file?

**Comment by lmsalgado** [comment]
> First, which version of Consul Democracy are you using?

The latest, I just today reinstalled from directly form github.com/consuldemocracy/installer


> Second, could you paste the contents of your delayed_job.log file?
It goes long now, but here's some relevant parts:

This are the normal entries, whe I try to send an email and it is not sent:
`I, [2024-12-09T06:11:21.874893 #34718]  INFO -- : 2024-12-09T06:11:21+0000: [Worker(delayed_job.0 host:pt-vrt-10 pid:34718)] Starting job worker
I, [2024-12-09T06:11:21.962198 #34728]  INFO -- : 2024-12-09T06:11:21+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:34728)] Starting job worker
I, [2024-12-09T06:36:18.919184 #34728]  INFO -- : 2024-12-09T06:36:18+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:34728)] Exiting...
I, [2024-12-09T06:36:18.919891 #34718]  INFO -- : 2024-12-09T06:36:18+0000: [Worker(delayed_job.0 host:pt-vrt-10 pid:34718)] Exiting...
I, [2024-12-09T07:34:00.057785 #12991]  INFO -- : 2024-12-09T07:34:00+0000: [Worker(delayed_job.0 host:pt-vrt-10 pid:12991)] Starting job worker
I, [2024-12-09T07:34:00.137170 #13002]  INFO -- : 2024-12-09T07:34:00+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:13002)] Starting job worker
I, [2024-12-11T13:34:20.724334 #13002]  INFO -- : 2024-12-11T13:34:20+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:13002)] Exiting...
I, [2024-12-11T13:34:20.727475 #12991]  INFO -- : 2024-12-11T13:34:20+0000: [Worker(delayed_job.0 host:pt-vrt-10 pid:12991)] Exiting...`

A failure during a reinstall, because in config it assumes a noreply@my_domain email account, that does not exist:
`E, [2024-12-16T17:17:04.674673 #9066] ERROR -- : 2024-12-16T17:17:04+0000: [Worker(delayed_job.0 host:pt-vrt-10 pid:9066)] Job ActionMailer::MailDeliveryJob [79235346-5e5c-470f-b47c-ea201b8135a3] from DelayedJob(default) with arguments: ["DeviseMailer", "confirmation_instructions", "deliver_now", {"args"=>[{"_aj_globalid"=>"gid://consul/User/3"}, "CKGxrkEFRoYWB8nvusSV", {"_aj_symbol_keys"=>[]}], "_aj_ruby2_keywords"=>["args"]}] (id=54) (queue=default) FAILED (2 prior attempts) with Net::SMTPFatalError: 573 noreply@my_domain.org unknown user account`

The lastest entries (after latest reinstalling and it started working for mails in queue and new ones... until I reboot):
`I, [2024-12-18T11:20:11.138175 #9086]  INFO -- : 2024-12-18T11:20:11+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:9086)] Job ActionMailer::MailDeliveryJob [d11e91a2-584a-42f3-ba62-94cd588a264d] from DelayedJob(default) with arguments: ["DeviseMailer", "confirmation_instructions", "deliver_now", {"args"=>[{"_aj_globalid"=>"gid://consul/User/2"}, "FFNcM71G-Fiyrkr91niP", {"_aj_symbol_keys"=>[]}], "_aj_ruby2_keywords"=>["args"]}] (id=77) (queue=default) COMPLETED after 1.3904
I, [2024-12-18T11:20:11.139948 #9086]  INFO -- : 2024-12-18T11:20:11+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:9086)] 4 jobs processed at 0.6912 j/s, 0 failed
I, [2024-12-18T11:20:11.333418 #9075]  INFO -- : 2024-12-18T11:20:11+0000: [Worker(delayed_job.0 host:pt-vrt-10 pid:9075)] Job ActionMailer::MailDeliveryJob [1f9ad166-3553-4f98-9afd-03392a2fc8cd] from DelayedJob(default) with arguments: ["DeviseMailer", "confirmation_instructions", "deliver_now", {"args"=>[{"_aj_globalid"=>"gid://consul/User/2"}, "FFNcM71G-Fiyrkr91niP", {"_aj_symbol_keys"=>[]}], "_aj_ruby2_keywords"=>["args"]}] (id=78) (queue=default) COMPLETED after 1.5328
I, [2024-12-18T11:20:11.335131 #9075]  INFO -- : 2024-12-18T11:20:11+0000: [Worker(delayed_job.0 host:pt-vrt-10 pid:9075)] 4 jobs processed at 0.6578 j/s, 0 failed
I, [2024-12-18T11:24:27.630652 #9086]  INFO -- : 2024-12-18T11:24:27+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:9086)] Job ActionMailer::MailDeliveryJob [c4140a29-80cd-4166-8e4a-e2548e2485c2] from DelayedJob(default) with arguments: ["DeviseMailer", "confirmation_instructions", "deliver_now", {"args"=>[{"_aj_globalid"=>"gid://consul/User/2"}, "FFNcM71G-Fiyrkr91niP", {"_aj_symbol_keys"=>[]}], "_aj_ruby2_keywords"=>["args"]}] (id=79) (queue=default) RUNNING
I, [2024-12-18T11:24:29.333960 #9086]  INFO -- : 2024-12-18T11:24:29+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:9086)] Job ActionMailer::MailDeliveryJob [c4140a29-80cd-4166-8e4a-e2548e2485c2] from DelayedJob(default) with arguments: ["DeviseMailer", "confirmation_instructions", "deliver_now", {"args"=>[{"_aj_globalid"=>"gid://consul/User/2"}, "FFNcM71G-Fiyrkr91niP", {"_aj_symbol_keys"=>[]}], "_aj_ruby2_keywords"=>["args"]}] (id=79) (queue=default) COMPLETED after 1.7032
I, [2024-12-18T11:24:29.335610 #9086]  INFO -- : 2024-12-18T11:24:29+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:9086)] 1 jobs processed at 0.5854 j/s, 0 failed
I, [2024-12-18T11:36:10.719119 #9086]  INFO -- : 2024-12-18T11:36:10+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:9086)] Job ActionMailer::MailDeliveryJob [de79a36c-fe62-41ff-a25a-7ba150437c48] from DelayedJob(default) with arguments: ["DeviseMailer", "confirmation_instructions", "deliver_now", {"args"=>[{"_aj_globalid"=>"gid://consul/User/3"}, "CKGxrkEFRoYWB8nvusSV", {"_aj_symbol_keys"=>[]}], "_aj_ruby2_keywords"=>["args"]}] (id=80) (queue=default) RUNNING
I, [2024-12-18T11:36:12.299414 #9086]  INFO -- : 2024-12-18T11:36:12+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:9086)] Job ActionMailer::MailDeliveryJob [de79a36c-fe62-41ff-a25a-7ba150437c48] from DelayedJob(default) with arguments: ["DeviseMailer", "confirmation_instructions", "deliver_now", {"args"=>[{"_aj_globalid"=>"gid://consul/User/3"}, "CKGxrkEFRoYWB8nvusSV", {"_aj_symbol_keys"=>[]}], "_aj_ruby2_keywords"=>["args"]}] (id=80) (queue=default) COMPLETED after 1.5802
I, [2024-12-18T11:36:12.301463 #9086]  INFO -- : 2024-12-18T11:36:12+0000: [Worker(delayed_job.1 host:pt-vrt-10 pid:9086)] 1 jobs processed at 0.6308 j/s, 0 failed`

**Reply by javierm** [reply]
To be honest, right now I'm a bit puzzled regarding why rebooting the computer results in a `noreply@my_domain.org unknown user account` error :thinking:.

Could you check whether, in the job trying to send the email, that address is the `FROM` address or the `TO` address?

**Reply by javierm** [reply]
Also is `my_domain.org` literally what's written in the error, or is it something you're using as a placeholder instead of your domain?

**Reply by lmsalgado** [reply]
I am so sorry, all my message caused a confusion. The noreply@my_domain.org comes up when I reinstall Consul Democracy, with command;
> ansible-playbook -v consul.yml -i hosts

Also, 

> my_domain.org

 is a placemark.

**Reply by javierm** [reply]
Hi, @lmsalgado :smile:. Sorry for the late reply :pray:.

Sorry, I misread your previous message and read "reinstall" as "reboot". The `ansible-playbook` command is not idempotent, so don't run it again if your Consul Democracy installation is already working. I guess that, after reinstalling, the `config/secrets.yml` file is overwritten and that's why you get the `unknown user account` error.

**Reply by lmsalgado** [reply]
Hi @javierm. Yes your are right, the reinstalling changed secrets.yml and temporally caused that specific error which I corrected immediately. Although my problem still persists, the cause is not that one.  Let us close this line of replies, to avoid added confusion. I will add more news to the last comment.

**Comment by lmsalgado** [comment]
@javierm : Thanks to your help, and the understanding that I should run the commands using the deploy user, I am getting close to the problem source.
As I said, after a fresh reinstall the emails work fine. Once I reboot the emails are not sent anymore. Now, If I run 
`RAILS_ENV=production bundle exec bin/delayed_job -m -n 2 restart`
the emails are back to be sent fine.

So, I guess that on reboot some service is not starting. Although I took care to note the services running before and after the reboot and they are the same. I used the command `service --status-all | grep '\[ + \]'`  (it may not be the more appropriated...).

Now, I noticed that crontab (in user Debug) now lists the command:
`@reboot /bin/bash -l -c 'cd /home/deploy/consul/releases/20241209060229 && RAILS_ENV=production bundle exec bin/delayed_job -m -n 2 restart'`
But the true is that once I reboot the emails stop again.



Update: I notice that [Consul version 2.0.0 ](https://github.com/consuldemocracy/consuldemocracy/releases/tag/2.0.0) had already implemented processes which monitor and restart DelayedJob processes in case they crash.  How do I check or configure that?  (using most recent version)

**Reply by javierm** [reply]
The processes themselves seem to start fine :thinking:. If there were problems with the processes monitoring the delayed jobs, the experience would be that they suddenly stop working (without a reboot). Using `htop`, you should see that there are two processes named `delayed_job.0_monitor` and `delayed_job.1_monitor`.

**Reply by lmsalgado** [reply]
Apparently the job either does not start on reboot, or it stops working right after reboot. My bet is that it does not start at all. Since I started the service manually, it has been 5 days running without stopping  (no reboots during that time, if I reboot the service  is sending the emails, even if crontab lists a task to start the service on reboot).

**Reply by javierm** [reply]
Thanks; will do more testing on production environments to see whether we can reproduce the issue when rebooting :thinking:.

**Reply by lmsalgado** [reply]
In resume, after all the messages before.

The problem still stands but we are getting close. Here's what I found until now:
The problem appears to be that the delayed_job job does not restart when the server is rebooted, or it fails right after. I still do not know why.

- If I run `ssh deply@....  `  and  `cd /home/deploy/consul/current && RAILS_ENV=production bundle exec bin/delayed_job -m -n 2 restart`  the process starts and all the pending emails are sent.

- I noticed that the config/schedule.rb file has a section
`every :reboot do
  command "cd #{@path} && RAILS_ENV=#{@environment} bundle exec bin/delayed_job -m -n 2 restart"
end`
Apparently this is not working, but I tried to add a section to force a job restart every hour:
`every :hour do
  command "cd /home/deploy/consul/current && RAILS_ENV=production bundle exec bin/delayed_job -m -n 2 restart"
end`
It still does not work, the job is never started as I still didn't receive the pending emails

- I tried a similar approach in crontab, the one on user deploy. It currently has the line:
`@reboot /bin/bash -l -c 'cd /home/deploy/consul/releases/20241209060229 && RAILS_ENV=production bundle exec bin/delayed_job -m -n 2 restart' `
I added the line:
`0 * * * * /bin/bash -l -c 'cd /home/deploy/consul/releases/20241209060229 && RAILS_ENV=production bundle exec bin/delayed_job -m -n 2 restart' `   
Still it does not work.

It seems that the only way to make it work is if I run the command manually (in the deploy user, it returns an error if I try to run it on root user as talked before).
I do not know why both schedule and crontab are failing to start/restart the job, neither how else I can automate a regular job restart to overcame this.
